ibis/expr/analysis.py

```diff
@@ -336,48 +336,29 @@ def propagate_down_window(func: ops.Value, frame: ops.WindowFrame):
     return type(func)(*clean_args)
 
 
-# TODO(kszucs): rewrite to receive and return an ops.Node
-def windowize_function(expr, frame):
-    assert isinstance(expr, ir.Expr), type(expr)
-    assert isinstance(frame, ops.WindowFrame)
-
-    def _windowize(op, frame):
-        if isinstance(op, ops.WindowFunction):
-            walked_child = _walk(op.func, frame)
-            walked = walked_child.to_expr().over(op.frame).op()
-        elif isinstance(op, ops.Value):
-            walked = _walk(op, frame)
-        else:
-            walked = op
-
-        if isinstance(walked, (ops.Analytic, ops.Reduction)):
-            return op.to_expr().over(frame).op()
-        elif isinstance(walked, ops.WindowFunction):
-            if frame is not None:
-                frame = walked.frame.copy(
-                    group_by=frame.group_by + walked.frame.group_by,
-                    order_by=frame.order_by + walked.frame.order_by,
-                )
-                return walked.to_expr().over(frame).op()
-            else:
-                return walked
-        else:
-            return walked
-
-    def _walk(op, frame):
-        # TODO(kszucs): rewrite to use the substitute utility
-        windowed_args = []
-        for arg in op.args:
-            if isinstance(arg, ops.Value):
-                arg = _windowize(arg, frame)
-            elif isinstance(arg, tuple):
-                arg = tuple(_windowize(x, frame) for x in arg)
-
-            windowed_args.append(arg)
-
-        return type(op)(*windowed_args)
+def windowize_function(expr, default_frame):
+    from ibis.common.patterns import Variable, _
+
+    func = Variable("func")
+    frame = Variable("frame")
+
+    node = expr.op()
+    res = node.replace(
+        (
+            p.WindowFunction(func, frame)
+            >> c.WindowFunction(
```

Comment:
This branch is only used from `GroupedTable` so it will better to remove this function entirely and use the specific replacement rules in `Projector` and `GroupedTable`.

Reply:
Well, we need both rewrites there as well.